File: CPI Silver Class/Week 1: Prefix Sums/Classwork/Problem: Max Subarray Sum/Main.java
import java.io.*;
import java.util.*;

public class Main {
    private static long solve(long[] nums) {
        long maxSum = nums[0];

        long minimumPrefixSum = 0;
        long currentPrefixSum = 0;

        for (int i = 0; i < nums.length; i++) {
            currentPrefixSum += nums[i];
            maxSum = Math.max(maxSum, currentPrefixSum - minimumPrefixSum);
            minimumPrefixSum = Math.min(minimumPrefixSum, currentPrefixSum);
        }

        return maxSum;
    }

    public static void main(String[] args) throws IOException {
        Kattio io = new Kattio();

        int n = io.nextInt();

        long[] nums = new long[n];
        for (int i = 0; i < n; i++) {
            nums[i] = io.nextLong();
        }

        io.println(solve(nums));

        io.close();
    }
}

class Kattio extends PrintWriter {
    private BufferedReader r;
    private StringTokenizer st;

    // standard input
    public Kattio() {
        this(System.in, System.out);
    }

    public Kattio(InputStream i, OutputStream o) {
        super(o);
        r = new BufferedReader(new InputStreamReader(i));
    }

    // USACO-style file input
    public Kattio(String problemName) throws IOException {
        super(new FileWriter(problemName + ".out"));
        r = new BufferedReader(new FileReader(problemName + ".in"));
    }

    // read next line
    public String readLine() {
        try {
            return r.readLine();
        } catch (IOException e) {
        }
        return null;
    }

    // returns null if no more input
    public String next() {
        try {
            while (st == null || !st.hasMoreTokens())
                st = new StringTokenizer(r.readLine());
            return st.nextToken();
        } catch (Exception e) {
        }
        return null;
    }

    public int nextInt() {
        return Integer.parseInt(next());
    }

    public double nextDouble() {
        return Double.parseDouble(next());
    }

    public long nextLong() {
        return Long.parseLong(next());
    }
}

File: CPI Silver Class/Week 1: Prefix Sums/Classwork/Problem: Max Subarray Sum/Main.cpp
#include <bits/stdc++.h>

using namespace std;

long solve(vector<long> nums) {
    long maxSum = nums[0];

    long minimumPrefixSum = 0;
    long currentPrefixSum = 0;

    for (int i = 0; i < nums.size(); i++) {
        currentPrefixSum += nums[i];
        maxSum = max(maxSum, currentPrefixSum - minimumPrefixSum);
        minimumPrefixSum = min(minimumPrefixSum, currentPrefixSum);
    }

    return maxSum;
}

int main() {
    int n;
    cin >> n;

    vector<long> nums(n);
    for (int i = 0; i < n; i++) {
        cin >> nums[i];
    }

    cout << solve(nums) << endl;

    return 0;

}
